detectors/gcp/cloud-function_test.go

```diff
@@ -0,0 +1,175 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"context"
+	"errors"
+	"os"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+)
+
+var (
+	errTest = errors.New("testError")
+)
+
+const (
+	projectIDValue = "some-projectID"
+	regionValue    = "some-region"
+	functionName   = "sample-function"
+)
+
+type metaDataClientImpl struct {
+	projectID  func() (string, error)
+	get        func(string) (string, error)
+	instanceID func() (string, error)
+}
+
+func (mock *metaDataClientImpl) ProjectID() (string, error) {
+	if mock.projectID != nil {
+		return mock.projectID()
+	}
+	return "", nil
+}
+
+func (mock *metaDataClientImpl) Get(key string) (string, error) {
+	if mock.get != nil {
+		return mock.get(key)
+	}
+	return "", nil
+}
+
+func (mock *metaDataClientImpl) InstanceID() (string, error) {
+	if mock.instanceID != nil {
+		return mock.instanceID()
+	}
+	return "", nil
+}
+
+type want struct {
+	res *resource.Resource
+	err error
+}
+
+func TestCloudFunctionDetect(t *testing.T) {
+	oldValue, ok := os.LookupEnv(gcpFunctionNameKey)
+	if !ok {
+		err := os.Setenv(gcpFunctionNameKey, functionName)
+		if err != nil {
+			t.Error("unable to set environment variable ", err)
+		}
+	}
+	defer func() {
+		if !ok {
+			os.Unsetenv(gcpFunctionNameKey)
+		} else {
+			os.Setenv(gcpFunctionNameKey, oldValue)
+		}
+	}()
```

Comment:
Use [`t.SetEnv()`](https://pkg.go.dev/testing#T.Setenv) instead.

```suggestion
	t.SetEnv(gcpFunctionNameKey, functionName)
```

Reply:
t.SetEnv(), is introduced in go1.17.
This may fail our ci actions, when using g1.16

https://pkg.go.dev/testing#T.Setenv

Reply:
You're right, I've been working on the collector too much! :)